File: src/main/java/net/ins/edu/algorithms/practice/BinarySearch.java
package net.ins.edu.algorithms.practice;

public class BinarySearch {

    private int search(int[] arr, int value) {
        if (arr.length == 0) return -1;
        return search(arr, value, 0, arr.length-1);
    }

    private int search(int[] arr, int value, int low, int high) {
        if (low > high) return -1;

        int mid = low + (high-low) / 2;

        if (arr[mid] == value) return mid;

        return value < arr[mid] ? search(arr, value, low, mid-1) : search(arr, value, mid+1, high);
    }

    public static void main(String[] args) {
        var arr = new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        for (int i = -1; i < arr.length + 1; i++) {
            System.out.println(new BinarySearch().search(arr, i)); // -1 for the first and the last elements
        }
    }
}

File: src/main/kotlin/net/ins/edu/algorithms/kt/search/BinarySearch.kt
package net.ins.edu.algorithms.kt.search

fun main() {
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 8))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 1))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 5))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 6))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 7))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 8))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000, 1001), 1000))
    println(binarySearch(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 11, 567, 1000), 1001))
}

fun binarySearch(arr: IntArray, search: Int, fromIndex: Int = 0, toIndex: Int = arr.size - 1): Int? {
    if (fromIndex <= toIndex) {
        val mid = fromIndex + (toIndex - fromIndex) / 2
        if (arr[mid] == search) return mid
        return if (search > arr[mid])
            binarySearch(arr, search, mid + 1, toIndex)
        else
            binarySearch(arr, search, fromIndex, mid - 1)
    }
    return -1
}